### PR TITLE
WIP: Refactor SqlDecorator.py

### PR DESCRIPTION
**Pull request by utkarsharma2**
Related: #279

As a part of our refactoring, we should create a class that only handles SQL execution that does not require any logic for a python decorator. This will make it easier to subclass SQL functions that are not decorators in the future.

Acceptance criteria:

- [ ] All tests should pass and there should be no change in top level functionality
- [ ] There should be a SQLExcecutor class that can execute SQL with schema handling but does not require a decorator
- [ ] The SQLDecorator class should be able to use the SQLExecutor class to reduce code duplicaiton.

### CODE REVIEW

File: src/astro/sql/operators/agnostic_aggregate_check.py [hunk 1]
```diff
@@ -53,19 +53,13 @@ def __init__(
 
         task_id = get_unique_task_id("aggregate_check")
 
-        def null_function():
-            pass
-
```
Comment (dimberman) [inline src/astro/sql/operators/agnostic_aggregate_check.py:58]: Huge +1, glad you found a way around this

File: src/astro/sql/operators/sql_decorator.py [hunk 2]
```diff
@@ -22,6 +26,127 @@
...
+        if context:
+            self.sql = self.render_template(self.sql, context)
+
+        # update parameters for snowflake table with fully qualified names
+        self._process_params()
+
+        cursor = self._run_sql(self.sql, self.parameters)
+        if self.handler is not None:
+            return self.handler(cursor)
+        return cursor
```
Comment (dimberman) [inline src/astro/sql/operators/sql_decorator.py:92]: @utkarsharma2 this seems to be missing a fair bit of logic (e.g. handling schemas, handling tables vs. temptables, etc.). Am I missing this somewhere?

File: src/astro/sql/operators/sql_decorator.py [hunk 3]
```diff
@@ -81,27 +213,41 @@ def hook(self):
...
         context = self._add_templates_to_context(context)
+
+        # render templating in sql query
         if context:
             self.sql = self.render_template(self.sql, context)
+
+        # update parameters for snowflake table with fully qualified names
         self._process_params()
 
         output_table_name: str = ""
```
Comment (dimberman) [inline src/astro/sql/operators/sql_decorator.py:253]: It seems like a decent amount of logic can be shared between the sql_decorator and sql_executor. Might be worth looking into how we can abstract more?

File: src/astro/utils/schema_util.py [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 
 def schema_exists(hook, schema, conn_type):
+    # why this is not happening for bigquery??
```
Comment (dimberman) [inline src/astro/utils/schema_util.py:8]: @tatiana can confirm but I seem to recall bigquery not having schemas?

File: src/astro/utils/table_handler.py [hunk 5]
```diff
@@ -15,6 +15,9 @@ def _set_variables_from_first_table(self):
         to create default values.
         """
         first_table: Optional[Table] = None
+
+        # where is op_args getting init from?
```
Comment (dimberman) [inline src/astro/utils/table_handler.py:19]: I'm not sure I understand this question?

**Comment by dimberman** [comment]
@utkarsharma2 can you please provide more description in the PR description?

**Comment by kaxil** [comment]
Since https://github.com/astro-projects/astro/pull/317 is merged, there are some merge conflicts

**Comment by kaxil** [comment]
closed in favor of https://github.com/astro-projects/astro/pull/353

**Pull request closed without merging**